src/pkcs11_uri_wrapped.rs

```diff
@@ -0,0 +1,229 @@
+    // pin(), which may still fail getting the PIN from a file for example.
+    pub fn has_pin(&self) -> bool {
+        match &self.p11uri.query_attributes.pin_value {
+            Some(_x) => return true,
+            None => {},
+        }
+        match &self.p11uri.query_attributes.pin_source {
+            Some(_x) => return true,
+            None => {},
+        }
+        false
+    }
+
+
+    // Get the module to use or an error in case no module could be found.
+    // First the module-path is checked for whether it holds an absolute that
+    // can be read by the current user. If this is the case the module is
+    // returned. Otherwise either the module-path is used or the user-provided
+    // module path is used to match a module containing what is set in the
+    // attribute module-name.
+    pub fn module(&self) -> Result<String> {
+        let searchdirs_tmp: Vec<String>; // FIXME
+        let searchdirs: &Vec<String>;
+
+        if let Some(mp) = &self.p11uri.query_attributes.module_path {
+            let info = std::fs::metadata(mp)?;
+            if info.is_file() {
+                // it's a file
+                if self.is_allowed_path(&mp, &self.allowed_module_paths)? {
+                    return Ok(mp.to_string());
+                }
+                return Err(anyhow!(""));
+            }
+            if !info.is_dir() {
+                // it's a symlink
+                return Err(anyhow!(""));
+            }
+            // it's a dir
+            searchdirs_tmp = vec![mp.to_string()];
+            searchdirs = &searchdirs_tmp;
+        } else {
+            searchdirs = &self.module_directories;
+        }
+
+        let module_name = match &self.p11uri.query_attributes.module_name {
+            Some(mn) => mn.to_lowercase(),
+            None => return Err(anyhow!("")),
+        };
+
+        for dir in searchdirs {
+            let file_results = match std::fs::read_dir(dir) {
+                Ok(fr) => fr,
+                Err(e) => continue,
+            };
+            for file_result in file_results {
+                let file = match file_result {
+                    Ok(f) => f,
+                    Err(e) => continue,
+                };
+                let file_lower = match file.file_name().into_string() {
+                    Ok(f) => f.to_lowercase(),
+                    Err(e) => continue,
+                };
+                let idx = match file_lower.find(&module_name) {
+                    Some(i) => i,
+                    None => continue,
+                };
+                // We require that file_lower ends with module_name or that
+                // a suffix follows so that softhsm will not match
+                // libsofthsm2.so but only libsofthsm.so
+                // FIXME: replace with idiomatic rust
+                if file_lower.len() == idx+module_name.len()
+                || file_lower.as_bytes()[idx+module_name.len()] == b'.' {
```

Comment:
try to use std::path::Path methods here to check if a file is inside a dir or something like that.